fix compile failure for getter with return lifetime of self

davidhewitt opened the pull request:
While testing `main` downstream I noticed a compile failure caused by #3318, when the value returned from a `#[getter]` has the same lifetime as the `&self` receiver.

This PR adds a corresponding test and fixes it. The solution is the usual trend for the macro function bodies towards expressions; there is a sequence of statements along the lines of `let item = ...; return convert(item)`, the PR changes this to just be `convert(...)`.

This doesn't need a changelog entry as the compile failure never hit a released version.

adamreichold commented on pyo3-macros-backend/src/pymethod.rs:
```diff
@@ -634,31 +634,28 @@ pub fn impl_py_getter_def(
-                ::std::result::Result::Ok(_pyo3::conversion::IntoPyPointer::into_ptr(item))
+                ::std::result::Result::Ok(
+                    _pyo3::conversion::IntoPyPointer::into_ptr(
+                        _pyo3::IntoPy::<_pyo3::Py<_pyo3::PyAny>>::into_py(::std::clone::Clone::clone(&(#slf.#field_token)), _py)
```
To be honest, I think this somewhat showcases the downside of the expression style insofar this expression is becoming an unreadable wall of text. The previous version with the extra binding was not great, but still more readable IMHO.

Could we at least try to use interpolation and multiple `quote!` invocation to try to pull this apart in our macro implementation while still emitting a single expression in the end?

davidhewitt replied:
Sure thing, I'll push a refactoring later (might create a bit of a bigger diff).

davidhewitt commented on pyo3-macros-backend/src/quotes.rs:
```diff
@@ -0,0 +1,15 @@
+    }
+}
+
+pub(crate) fn map_result_into_ptr(result: TokenStream) -> TokenStream {
+    quote! {
+        #result.map(_pyo3::IntoPyPointer::into_ptr)
+    }
+}
```
I think this function can be removed (and code moved out of the macros) by moving the `Py<PyAny> -> *mut ffi::PyObject` conversion into the trampolines. That'll also have a positive effect on code size, so I'll enact that as a follow-up within a few days.

davidhewitt reviewed (commented):
Ok, I've pushed a commit which improves the implementation of `impl_py_getter_def` a little bit.

I've identified a follow-up which will move code out of the macro-generated code and into the trampolines, so I'd like to ask to merge this as-is (even if the macro expression is still a bit verbose) because I'll simplify further in the follow up.

davidhewitt commented:
I also pushed a commit which moves all usage of the `_py` ident to `py`, which allowed a general reduction in complexity across the macro code (including inside this `impl_py_getter_def` once I moved to the `ok_wrap` form).

adamreichold commented on pyo3-macros-backend/src/method.rs:
```diff
@@ -164,19 +167,15 @@ impl SelfType {
-                )
+                error_mode.handle_error(quote_spanned! { *span =>
+                    _pyo3::impl_::extract_argument::#method::<#cls>(
+                        #py.from_borrowed_ptr::<_pyo3::PyAny>(#slf),
```
Below, `py` is used, but here it stays `#py`. Oversight or deliberate?

davidhewitt replied:
Most likely oversight 🙈

adamreichold commented on pyo3-macros-backend/src/method.rs:
```diff
@@ -155,7 +156,9 @@ impl ExtractErrorMode {
 
 impl SelfType {
     pub fn receiver(&self, cls: &syn::Type, error_mode: ExtractErrorMode) -> TokenStream {
-        let py = syn::Ident::new("_py", Span::call_site());
```
Do we know that there was no reason to do this and/or that we have a regression test that which catch any issues introduce by not quoting the GIL token?

davidhewitt replied:
Let's move this to #3356

adamreichold reviewed (commented):
While I welcome the reduction in Perl-style sigils by not quoting the GIL token, I somewhat fear regressions in macro hygiene due to this.

I am also not sure what the motivation for doing that change now is? Sure, this implies another reduction in sigil noise in the proc macro code, but to me, it appears small compared to possible regression in macro hygiene, especially which such a common identifier as `py` (at least in PyO3 code bases).

I am pretty sure I am missing something, because this appears somewhat risky to me.

davidhewitt commented:
That's a reasonable concern. The `py` ident should always have `call_site` hygiene, and the only case that it's necessary to explicitly construct the ident is when mixing with `quote_spanned` to emit other tokens with a different hygiene. (The interpolated `#py` in a `quote_spanned` will keep its original span.)

The `py` ident was parameterised further in some functions because we have used both `py` and `_py` as the identifier inside the macros codebase.

To remove the hygiene concerns from this PR I've moved that commit to #3356 and replaced it here with a simpler commit which unifies all idents as `py` in the macros codebase. That's sufficient to allow the `ok_wrap` helper to work without needing to take `py` ident as an argument.
